### PR TITLE
[diffusion] model: support LTX2.3 high quality pipeline

### PR DESCRIPTION
**Pull request by mickqian**
psnr: 20.71 dB

sglang:

https://github.com/user-attachments/assets/49bbdaa8-1aec-4f91-91a8-b615c98d167a

official:

https://github.com/user-attachments/assets/68209770-375a-4f38-b360-6e6a435abff8

## Reproduce
  
prompt:
```
00:00 - 00:02.5
Visual / Action:
SpongeBob stands in front of a mirror in his living room. He is wearing a brand new jellyfish-shaped hat with two glowing jellyfish eyes on top. He turns his head left and right, admiring himself with a big proud smile.

Dialogue:
SpongeBob (to himself, cheerful): "Perfect!"

SFX:
Upbeat whistling tune (light and bouncy), soft electric buzz from the hat's glowing eyes.

00:02.5 - 00:05
Visual / Action:
Suddenly, Patrick jumps out from behind SpongeBob, clapping his hands loudly and shouting. SpongeBob gets so scared that he jumps straight up, and the jellyfish hat flies off his head into the air.

Dialogue:
Patrick (loud and blunt): "Ugly hat!"

SFX:
Short dramatic sting - "DAH!" (comedy suspense), followed by a "whoosh" sound as the hat flies upward.

00:05 - 00:07.5
Visual / Action:
The hat falls down and lands perfectly on Patrick's head. Immediately, the hat's jellyfish eyes turn from yellow to red, and it makes an angry buzzing sound. Then, with a soft "poof", the hat sprays a small blob of pink goo directly onto Patrick's face.

Dialogue:
Patrick (confused, muffled): "Does it... hate me?"

SFX:
Warning buzz (low to high pitch), followed by a wet "plop" / "pfft" of the goo spraying, and a comedic slide whistle descending.

00:07.5 - 00:10
Visual / Action:
Patrick sticks out his tongue and licks the pink goo off his cheek. His eyes suddenly light up with joy, and he grins widely. SpongeBob doubles over, holding his stomach and laughing.

Dialogue:
Patrick (excited, mouth half-full): "It's strawberry flavored! Do it again!"

SFX:
Licking sound "lap", then a cheerful ukulele jingle (4 quick notes) + a final "ding!"
```

```bash
sglang generate \
  --model-path Lightricks/LTX-2.3 \
  --pipeline-class-name LTX2TwoStageHQPipeline \
  --num-gpus 1 \
  --ltx2-two-stage-device-mode original \
  --text-encoder-cpu-offload \
  --dit-layerwise-offload true \
  --width 768 \
  --height 512 \
  --num-frames 241 \
  --fps 24 \
  --num-inference-steps 30 \
  --prompt "$(...)"
 ```

official:
```python

from __future__ import annotations

import argparse
from pathlib import Path

import torch

try:
    from ltx_core.components.guiders import MultiModalGuiderParams
    from ltx_core.loader import LTXV_LORA_COMFY_RENAMING_MAP, LoraPathStrengthAndSDOps
    from ltx_core.loader.single_gpu_model_builder import SingleGPUModelBuilder
    from ltx_core.model.video_vae import TilingConfig, get_video_chunks_number
    from ltx_pipelines.ti2vid_two_stages_hq import TI2VidTwoStagesHQPipeline
    from ltx_pipelines.utils.constants import DEFAULT_NEGATIVE_PROMPT
    from ltx_pipelines.utils.media_io import encode_video
    from transformers.models.gemma3.modeling_gemma3 import (
        Gemma3RotaryEmbedding,
        Gemma3TextScaledWordEmbedding,
    )
    from transformers.models.siglip.modeling_siglip import SiglipVisionEmbeddings
except ImportError as exc:  # pragma: no cover - manual helper
    raise SystemExit(
        "This script must run against the official LTX-2 repo. "
        "Set PYTHONPATH to include packages/ltx-core/src and packages/ltx-pipelines/src."
    ) from exc


PR23366_SPONGEBOB_PROMPT = """00:00 - 00:02.5
Visual / Action:
SpongeBob stands in front of a mirror in his living room. He is wearing a brand new jellyfish-shaped hat with two glowing jellyfish eyes on top. He turns his head left and right, admiring himself with a big proud smile.

Dialogue:
SpongeBob (to himself, cheerful): "Perfect!"

SFX:
Upbeat whistling tune (light and bouncy), soft electric buzz from the hat's glowing eyes.

00:02.5 - 00:05
Visual / Action:
Suddenly, Patrick jumps out from behind SpongeBob, clapping his hands loudly and shouting. SpongeBob gets so scared that he jumps straight up, and the jellyfish hat flies off his head into the air.

Dialogue:
Patrick (loud and blunt): "Ugly hat!"

SFX:
Short dramatic sting - "DAH!" (comedy suspense), followed by a "whoosh" sound as the hat flies upward.

00:05 - 00:07.5
Visual / Action:
The hat falls down and lands perfectly on Patrick's head. Immediately, the hat's jellyfish eyes turn from yellow to red, and it makes an angry buzzing sound. Then, with a soft "poof", the hat sprays a small blob of pink goo directly onto Patrick's face.

Dialogue:
Patrick (confused, muffled): "Does it... hate me?"

SFX:
Warning buzz (low to high pitch), followed by a wet "plop" / "pfft" of the goo spraying, and a comedic slide whistle descending.

00:07.5 - 00:10
Visual / Action:
Patrick sticks out his tongue and licks the pink goo off his cheek. His eyes suddenly light up with joy, and he grins widely. SpongeBob doubles over, holding his stomach and laughing.

Dialogue:
Patrick (excited, mouth half-full): "It's strawberry flavored! Do it again!"

SFX:
Licking sound "lap", then a cheerful ukulele jingle (4 quick notes) + a final "ding!"
"""

WIDTH = 768
HEIGHT = 512
NUM_FRAMES = 241
FRAME_RATE = 24.0
NUM_INFERENCE_STEPS = 30
DISTILLED_LORA_STAGE_1 = 0.25
DISTILLED_LORA_STAGE_2 = 0.5

VIDEO_GUIDER_PARAMS = MultiModalGuiderParams(
    cfg_scale=3.0,
    stg_scale=0.0,
    rescale_scale=0.45,
    modality_scale=3.0,
    skip_step=0,
    stg_blocks=[],
)
AUDIO_GUIDER_PARAMS = MultiModalGuiderParams(
    cfg_scale=7.0,
    stg_scale=0.0,
    rescale_scale=1.0,
    modality_scale=3.0,
    skip_step=0,
    stg_blocks=[],
)


def _materialize_known_meta_buffers(model: torch.nn.Module, device: torch.device) -> None:
    module_map = dict(model.named_modules())

    for full_name, buffer in model.named_buffers():
        if str(buffer.device) != "meta":
            continue

        module_name, buffer_name = full_name.rsplit(".", 1)
        module = module_map[module_name]

        if isinstance(module, SiglipVisionEmbeddings) and buffer_name == "position_ids":
            module._buffers[buffer_name] = torch.arange(
                module.num_positions, device=device, dtype=torch.long
            ).expand((1, -1))
            continue

        if isinstance(module, Gemma3TextScaledWordEmbedding) and buffer_name == "embed_scale":
            module._buffers[buffer_name] = torch.tensor(
                module.embedding_dim**0.5, device=device
            )
            continue

        if isinstance(module, Gemma3RotaryEmbedding) and buffer_name == "inv_freq":
            inv_freq, attention_scaling = module.rope_init_fn(module.config, device)
            module._buffers[buffer_name] = inv_freq
            module.original_inv_freq = inv_freq
            module.attention_scaling = attention_scaling


def _patch_official_builder_for_gemma_buffers() -> None:
    if getattr(SingleGPUModelBuilder, "_pr23366_meta_patch_applied", False):
        return

    original_return_model = SingleGPUModelBuilder._return_model

    def _patched_return_model(self, meta_model: torch.nn.Module, device: torch.device):
        _materialize_known_meta_buffers(meta_model, device)
        return original_return_model(self, meta_model, device)

    SingleGPUModelBuilder._return_model = _patched_return_model
    SingleGPUModelBuilder._pr23366_meta_patch_applied = True


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Pinned official LTX-2.3 HQ two-stage reproduction for the SpongeBob "
            "storyboard case used in SGLang PR #23366."
        )
    )
    parser.add_argument("--checkpoint-path", required=True)
    parser.add_argument("--distilled-lora-path", required=True)
    parser.add_argument("--spatial-upsampler-path", required=True)
    parser.add_argument("--gemma-root", required=True)
    parser.add_argument("--output-path", required=True)
    parser.add_argument(
        "--seed",
        type=int,
        default=10,
        help=(
            "Pinned seed for stable reference generation. "
            "PR #23366 did not spell out a seed in the sample command, but this "
            "script defaults to 10 so the official reference is reproducible."
        ),
    )
    parser.add_argument(
        "--negative-prompt",
        default=DEFAULT_NEGATIVE_PROMPT,
        help="Override the official default negative prompt if needed.",
    )
    return parser.parse_args()


def _clone_video_chunks(chunks):
    for chunk in chunks:
        yield chunk.clone()


@torch.inference_mode()
def main() -> None:
    args = parse_args()
    _patch_official_builder_for_gemma_buffers()
    output_path = Path(args.output_path).expanduser().resolve()
    output_path.parent.mkdir(parents=True, exist_ok=True)

    pipeline = TI2VidTwoStagesHQPipeline(
        checkpoint_path=str(Path(args.checkpoint_path).expanduser().resolve()),
        distilled_lora=[
            LoraPathStrengthAndSDOps(
                str(Path(args.distilled_lora_path).expanduser().resolve()),
                1.0,
                LTXV_LORA_COMFY_RENAMING_MAP,
            )
        ],
        distilled_lora_strength_stage_1=DISTILLED_LORA_STAGE_1,
        distilled_lora_strength_stage_2=DISTILLED_LORA_STAGE_2,
        spatial_upsampler_path=str(
            Path(args.spatial_upsampler_path).expanduser().resolve()
        ),
        gemma_root=str(Path(args.gemma_root).expanduser().resolve()),
        loras=(),
    )

    tiling_config = TilingConfig.default()
    video, audio = pipeline(
        prompt=PR23366_SPONGEBOB_PROMPT,
        negative_prompt=args.negative_prompt,
        seed=args.seed,
        height=HEIGHT,
        width=WIDTH,
        num_frames=NUM_FRAMES,
        frame_rate=FRAME_RATE,
        num_inference_steps=NUM_INFERENCE_STEPS,
        video_guider_params=VIDEO_GUIDER_PARAMS,
        audio_guider_params=AUDIO_GUIDER_PARAMS,
        images=[],
        tiling_config=tiling_config,
    )

    encode_video(
        video=_clone_video_chunks(video),
        fps=FRAME_RATE,
        audio=audio,
        output_path=str(output_path),
        video_chunks_number=get_video_chunks_number(NUM_FRAMES, tiling_config),
    )
    print(output_path)


if __name__ == "__main__":
    main()
```
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Review by nidhishgajjar — CHANGES_REQUESTED** [review]
This PR is still in draft status with a template description only. To enable proper review, please:

1. **Fill in the description template** with actual details about:
   - Motivation: Why is this change needed?
   - Modifications: What files were changed and what do they do?
   - Accuracy Tests: Provide test results if this affects model outputs
   - Speed Tests: Provide benchmarking if this affects inference speed

2. **Complete the checklist** items:
   - Format code with pre-commit
   - Add unit tests
   - Update documentation
   - Provide accuracy and speed benchmarks

3. **Remove draft status** once ready for review

This is a substantial change (4416 additions, 24 files) for LTX2.3 high quality pipeline support, so proper documentation and testing information is essential before it can be reviewed.
